sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestCases/ConditionEvaluatesToConstant.cs

```diff
@@ -1530,33 +1530,203 @@ class FooContainer
+
+      if (result == null || !result.Succeed)
+      {
+        Console.WriteLine("shorted");
+        if (result != null)
+        {
+          Console.WriteLine("other");
+        }
+      }
+
+      if (result?.Succeed != true)
+      {
+        Console.WriteLine("shorted");
+        if (result != null)
+        {
+          Console.WriteLine("other");
+        }
+      }
+    }
+
+    public static void NonCompliant1()
+    {
+      Result result = null;
+      if (result?.Succeed != null) // Noncompliant
+      { // Secondary
+        Console.WriteLine("shorted");
+        if (result != null)
+        {
+          Console.WriteLine("other");
+        }
+      }
+    }
+
+    public static void NonCompliant2()
+    {
+      Result result = new Result();
+      if (result?.Succeed != null)
+      {
+        Console.WriteLine("shorted");
+        while (result != null) // Noncompliant
+        {
+          Console.WriteLine("other");
+        }
+      }
+    }
+
+    public class A
+    {
+      public bool booleanVal { get; set; }
+    }
+
+    public static void Compliant2()
+    {
+      A aObj = null;
+      if (aObj?.booleanVal ?? false)
+      {
+        Console.WriteLine("a");
+      }
+    }
+
+    public static void NonCompliant3()
+    {
+      A aObj = null;
+      if (aObj?.booleanVal == null) // Noncompliant
+      {
+        Console.WriteLine("a");
+      }
+
+      if (aObj?.booleanVal != null) // Noncompliant
+      { // Secondary
+        Console.WriteLine("a");
+      }
+    }
+
+    public static void Compliant3(A a)
+    {
+
+      if (a?.booleanVal == true)
+      {
+        Console.WriteLine("Compliant");
+        return;
+      }
+
+      if (a != null)  // Compliant
+      {
+      }
+    }
+
+    public static void NonCompliant4(A a)
+    {
+
+      if (a?.booleanVal == null)
+      {
+        Console.WriteLine("Compliant");
+        return;
+      }
+
+      if (a != null) // Noncompliant
+      {
+      }
+    }
+
+    public static void Compliant4(A a)
+    {
+      if (a?.booleanVal == null)
+      {
+        Console.WriteLine("Compliant");
+      }
+
+      if (a != null) // Compliant
+      {
+      }
+    }
+
+    public static void Compliant5(A a)
+    {
+      while (a?.booleanVal == null ? true : false)
+      {
+        Console.WriteLine("Compliant");
+      }
+    }
+
+    public static void NonCompliant5()
+    {
+      A a = null;
+      while (a?.booleanVal == null ? true : false) // Noncompliant
+      { // Secondary@-1
+        Console.WriteLine("Compliant");
+      }
+    }
+
+    public class S
+    {
+      public string str=null;
+    }
+
+    public static void Compliant6(S sObj)
+    {     
+      if (sObj?.str?.Length > 2)
+      {
+        Console.WriteLine("a");
+      }
+    }
+
+    public static void NonCompliant6()
+    {
+      S sObj = null; 
+      if (sObj?.str?.Length > 2) // Noncompliant
+      { // Secondary
+        Console.WriteLine("a");
+      }
+    }
+  }
+
```

Comment:
you could also add tests with `is` inside this file, right?

like 
```
 if ((control.Tag is string i) && !string.IsNullOrEmpty(i)) // <= S2589 FP
```

(with `IsNull` and with `IsNullOrEmpty` as wel)...